### PR TITLE
ap-zkml-article

### PR DESCRIPTION
**Pull request by AndrePanin**

### CODE REVIEW

File: docs/examples/zkml.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+This article presents a proof-of-concept implementation of a zero-knowledge proof application designed for Gear-powered blockchains. You can access the source code on [GitHub](https://github.com/gear-foundation/external-zkml-actor).
+
+:::note
+Keep in mind that this example is intended for educational purposes as a technology demonstration. It may contain inaccuracies, suboptimal optimizations, and security vulnerabilities. The solutions used are not recommended for production use.
+:::
+
+### How It Works
+
+This implementation offers the capability to shift resource-intensive computations away from the blockchain while ensuring their integrity. It demonstrates the evaluation of the MNIST machine learning model as an illustrative example. This is accomplished by generating zero-knowledge proofs of computations and subsequently verifying them on-chain.
```
Comment (mertwole) [inline docs/examples/zkml.md:38]: wikipedia link here? https://en.wikipedia.org/wiki/MNIST_database

File: docs/examples/zkml.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+The associated smart contract allows only one verification session per client. If a client wishes to discard the current verification and submit new inputs, they have the option to execute a purge verification.
```
Comment (mertwole) [inline docs/examples/zkml.md:40]:
```suggestion
The associated smart contract allows only one verification session per client. If a client wishes to discard the current verification and submit new inputs, they have the option to send a transaction to purge in-progress verification.
```

File: docs/examples/zkml.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+1. `Initializer`: This participant is responsible for deploying the `zk-smart-contract` onto the blockchain and initializing its parameters. During initialization, the structure of the machine learning model is defined, setting the stage for subsequent interactions.
+
+2. `User`: Users of the system interact with the `zk-smart-contract` by submitting their input data, which is intended for processing by the machine learning model. This step demonstrates how external parties can seamlessly engage with the blockchain without directly executing computationally intensive operations.
+
+3. `Prover`: The Prover is deployed on a server and continuously listens for incoming events, computes outputs, and proves computations. The Prover retrieves input from the `zk-smart-contract`, evaluates the machine learning model, and generates a zero-knowledge proof that attests to the validity of the computation. The resulting output of the machine learning model, along with the zero-knowledge proof, is then submitted back to the `zk-smart-contract`.
```
Comment (mertwole) [inline docs/examples/zkml.md:52]:
```suggestion
3. `Prover`: The Prover is deployed on a server and continuously polling `zk-smart-contract` to discover new inputs provided by `User`s, computes outputs, and proves computations. The Prover retrieves input from the `zk-smart-contract`, evaluates the machine learning model, and generates a zero-knowledge proof that attests to the validity of the computation. The resulting output of the machine learning model, along with the zero-knowledge proof, is then submitted back to the `zk-smart-contract`.
```

File: docs/examples/img/zkml.jpg
Comment (mertwole) [inline docs/examples/img/zkml.jpg:1]: I think 5th step can be removed from scheme, as it's too low-level details of implementation

Comment (mertwole) [reply docs/examples/img/zkml.jpg:1]: And also some details about data formats used can be omitted too:
- input: grayscale image ~resolution = ...~
- output: ten ~128-bit~ numbers ...
- ~proof: byte array~ proof data

Comment (mertwole) [reply docs/examples/img/zkml.jpg:1]: Client may be another smart-contract, but may be not (also `User` is used everywhere, but `Client` on scheme)

Comment (mertwole) [reply docs/examples/img/zkml.jpg:1]: Also would be nice to mark somewhere that the bottom scheme is the `Prover` logics

File: docs/examples/zkml.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+# ZKML: Empowering Privacy on Vara Network
+
+## Introduction
+
+Imagine you have a secret way to prove something is true without revealing any extra details. That's precisely what a zero-knowledge (ZK) proof does in the digital world. It's like a special handshake between two parties: one party (the prover) convinces the other party (the verifier) that something is true without exposing anything else.
+
+In the world of Web3, zero-knowledge cryptography is a game-changer. It empowers developers to create scalable and private applications. This technology allows them to conceal complex calculations while ensuring their accuracy and to create proofs of computational integrity that are much easier to verify than to compute.
+
+Machine learning (ML) is a branch of artificial intelligence that enables computers to learn and improve from data, optimizing their performance through iterative processes.
```
[the remark this answers is not in the document]
Comment (AndrePanin) [reply docs/examples/zkml.md]: performance is not only about speed, it also about accuracy, but changed to be more precise.

File: docs/examples/zkml.md (hunk 1 above)
Comment (AndrePanin) [reply docs/examples/zkml.md]: added

File: docs/examples/img/zkml.jpg
Comment (AndrePanin) [reply docs/examples/img/zkml.jpg:1]: updated

**Pull request merged**